src/Plutus/Model/Contract.hs

```diff
@@ -272,16 +275,22 @@ spend pkh val = do
   mSp <- spend' pkh val
   pure $ fromJust mSp
 
-{- | User wants to spend money.
- It returns input UTXOs and output UTXOs for change.
- Note that it does not removes UTXOs from user account.
- We can only spend by submitting TXs, so if you run it twice
- it will choose from the same set of UTXOs.
+{- | Generate a transaction to spend value.
+
+Does not submit the transaction, or mark UTxOs as spent.
+
+Will not spend value from a UTxO if it contains a reference script, so as not to
+  make the script unreferencable.
 -}
 spend' :: PubKeyHash -> Value -> Run (Maybe UserSpend)
 spend' pkh expected = do
   refs <- txOutRefAt (pubKeyHashAddress pkh)
-  mUtxos <- (\m -> mapM (\r -> (r,) <$> M.lookup r m) refs) <$> gets mockUtxos
+  mUtxos <-
+    gets
+      ( (\m -> mapM (\r -> (r,) <$> M.lookup r m) refs)
+          . M.filter (isNothing . txOutReferenceScript)
```

Comment:
So it is still impossible to spend UTxO with reference script?

Reply:
`spend'` will not automatically find a UTxO with reference script because it can accidentally make the script unreferencable. Directly spending those UTxOs will work (I think it always did).

Reply:
Alright, makes sense